File: programming/lab3/src/com/armemius/lab3/Entity.java
package com.armemius.lab3;

import java.util.Objects;

public abstract class Entity implements ITalker, IMoving {
    public Entity(String name, Status status) {
        this.name = name;
        this.status = status;
    }

    private final String name;
    protected Status status;

    public String getName() {
        return this.name;
    }

    @Override
    public void moveTo(String name, MoveAction action, IMoving... companions) {
        if (companions.length == 0) {
            System.out.println(this + " идёт в: " + name);
            action.action();
            System.out.println(this + " пришёл");
        } else {
            String names = this.toString();
            for (int it = 0; it < companions.length; ++it) {
                if (it + 1 == companions.length) {
                    names += " и " + companions[it].toString();
                } else {
                    names += ", " + companions[it].toString();
                }
            }
            System.out.println(names + " идут в: " + name);
            action.action();
            System.out.println(names + " пришли");
        }

    }

    @Override
    public void tell(ITalker name, String text, ITalker... companions) {
        if (companions.length == 0) {
            System.out.println(this + " говорит " + name + ": " + text);
        } else {
            String names = this.toString();
            for (int it = 0; it < companions.length; ++it) {
                if (it + 1 == companions.length) {
                    names += " и " + companions[it].toString();
                } else {
                    names += ", " + companions[it].toString();
                }
            }
            System.out.println(names + " говорят " + name + ": " + text);
        }
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Entity entity = (Entity) o;
        return Objects.equals(name, entity.name) && status == entity.status;
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, status);
    }
}

File: programming/lab4/src/main/kotlin/com/armemius/lab4/entity/Entity.kt
package com.armemius.lab4.entity

import com.armemius.lab4.exception.MoveException
import com.armemius.lab4.location.Location

open class Entity(private val name: String): IMoving, ITalking, ISinger {
    private var location: Location = Location.getLocation("default")
    override fun moveTo(name: String, action: () -> Unit, vararg companions: IMoving) {
        if (name == location.name)
            throw MoveException("Trying to move to the same location from origin")
        location = Location.getLocation(name)
        if (companions.isEmpty()) {
            println("$this идёт в: ${location.name}");
            action()
            println("$this пришёл")
        } else {
            var names: String = this.toString();
            for (it in 0 until companions.size) {
                names += if (it + 1 == companions.size) {
                    " и ${companions[it]}"
                } else {
                    ", ${companions[it]}"
                }
            }
            println("$names идут в: ${location.name}");
            action();
            println("$names пришли");
        }

    }

    override fun tell(entity: ITalking?, message: String, vararg companions: ITalking) {
        if (entity == null) {
            println("$this говорит про себя: $message");
            return
        }
        if (companions.isEmpty()) {
            println("$this говорит name: $message");
        } else {
            var names: String = this.toString();
            for (it in 0 until companions.size) {
                names += if (it + 1 == companions.size) {
                    " и ${companions[it]}"
                } else {
                    ", ${companions[it]}"
                }
            }
            println("$names говорят $name: $message");
        }
    }

    override fun sing(name: String) {
        println("$this поёт $name")
    }

    override fun toString(): String {
        return name
    }
}

